src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs

```diff
@@ -48,238 +51,314 @@ public sealed class SqlServerFhirModel : IHostedService
-                SELECT * FROM  OPENJSON (@searchParams) 
-                WITH (Uri varchar(128) '$.Uri')
-                EXCEPT SELECT Uri FROM dbo.SearchParam;
-
-                -- result set 2
-                SELECT Uri, SearchParamId FROM dbo.SearchParam;
-
-                INSERT INTO dbo.ClaimType (Name) 
-                SELECT value FROM string_split(@claimTypes, ',')
-                EXCEPT SELECT Name FROM dbo.ClaimType; 
-
-                -- result set 3
-                SELECT ClaimTypeId, Name FROM dbo.ClaimType;
-
-                INSERT INTO dbo.CompartmentType (Name) 
-                SELECT value FROM string_split(@compartmentTypes, ',')
-                EXCEPT SELECT Name FROM dbo.CompartmentType; 
-
-                -- result set 4
-                SELECT CompartmentTypeId, Name FROM dbo.CompartmentType;
-                
-                COMMIT TRANSACTION
-
-                -- result set 5
-                SELECT Value, SystemId from dbo.System;
-
-                -- result set 6
-                SELECT Value, QuantityCodeId FROM dbo.QuantityCode";
-
-            string commaSeparatedResourceTypes = string.Join(",", ModelInfoProvider.GetResourceTypeNames());
-            string searchParametersJson = JsonConvert.SerializeObject(_searchParameterDefinitionManager.AllSearchParameters.Select(p => new { Name = p.Name, Uri = p.Url }));
-            string commaSeparatedClaimTypes = string.Join(',', _securityConfiguration.PrincipalClaims);
-            string commaSeparatedCompartmentTypes = string.Join(',', ModelInfoProvider.GetCompartmentTypeNames());
-
-            sqlCommand.Parameters.AddWithValue("@resourceTypes", commaSeparatedResourceTypes);
-            sqlCommand.Parameters.AddWithValue("@searchParams", searchParametersJson);
-            sqlCommand.Parameters.AddWithValue("@claimTypes", commaSeparatedClaimTypes);
-            sqlCommand.Parameters.AddWithValue("@compartmentTypes", commaSeparatedCompartmentTypes);
-
-            await using SqlDataReader reader = await sqlCommand.ExecuteReaderAsync(CommandBehavior.SequentialAccess, cancellationToken);
-
-            var resourceTypeToId = new Dictionary<string, short>(StringComparer.Ordinal);
-            var resourceTypeIdToTypeName = new Dictionary<short, string>();
-            var searchParamUriToId = new Dictionary<Uri, short>();
-            var systemToId = new ConcurrentDictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-            var quantityCodeToId = new ConcurrentDictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-            var claimNameToId = new Dictionary<string, byte>(StringComparer.Ordinal);
-            var compartmentTypeToId = new Dictionary<string, byte>();
-
-            // result set 1
-            while (await reader.ReadAsync(cancellationToken))
+            // If we are apply a full snap shot schema file, or if the server is just starting up
+            if (runAllInitialization || _highestInitializedVersion == 0)
```

Comment:
Ah, yes good question. I found this edge case recently while doing some testing locally. Here is the scenario:

1) Run the FHIR server on schema version 5 (I did this by setting `SchemaVersionConstants.Max = SchemaVersion.V5`)
2) Stop the FHIR server
3) Run the FHIR server with schema version 6 and automatic updates enabled

In this case, `runAllInitialization` will be `false` because we are applying a diff schema, but we want to call `InitializeBase()` to initialize the C# dictionaries. `InitializeBase()` won't be triggered on first request either because `_highestInitializedVersion` is already set to 6 at that point.

Reply:
Makes sense!